libsignal-protocol/src/stores/session_store.rs

```diff
@@ -0,0 +1,204 @@
+    ) -> Result<Vec<i32>, InternalError>;
+
+    /// Determine whether there is a committed session record for a
+    /// recipient ID + device ID tuple.
+    fn contains_session(&self, addr: Address) -> Result<bool, InternalError>;
+
+    /// Commit to storage the session record for a given recipient ID + device
+    /// ID tuple.
+    fn store_session(
+        &self,
+        addr: Address,
+        session: SerializedSession,
+    ) -> Result<(), InternalError>;
+
+    /// Remove a session record for a recipient ID + device ID tuple.
+    fn delete_session(&self, addr: Address) -> Result<(), InternalError>;
+
+    /// Remove the session records corresponding to all devices of a recipient
+    /// ID.
+    ///
+    /// Returns the number of deleted sessions.
+    fn delete_all_sessions(&self, name: &[u8]) -> Result<usize, InternalError>;
+}
+
+pub(crate) fn new_vtable<S: SessionStore + 'static>(
+    session_store: S,
+) -> sys::signal_protocol_session_store {
+    let state: Box<State> = Box::new(State(Box::new(session_store)));
+
+    sys::signal_protocol_session_store {
+        user_data: Box::into_raw(state) as *mut c_void,
+        load_session_func: Some(load_session_func),
+        get_sub_device_sessions_func: Some(get_sub_device_sessions_func),
+        store_session_func: Some(store_session_func),
+        contains_session_func: Some(contains_session_func),
+        delete_session_func: Some(delete_session_func),
+        delete_all_sessions_func: Some(delete_all_sessions_func),
+        destroy_func: Some(destroy_func),
+    }
+}
+
+struct State(Box<dyn SessionStore>);
+
+unsafe extern "C" fn load_session_func(
+    record: *mut *mut sys::signal_buffer,
+    user_record: *mut *mut sys::signal_buffer,
+    address: *const sys::signal_protocol_address,
+    user_data: *mut c_void,
+) -> c_int {
+    assert!(!record.is_null());
+    assert!(!user_record.is_null());
+    assert!(!address.is_null());
+    assert!(!user_data.is_null());
+
+    let state = &*(user_data as *const State);
+    let address = Address::from_ptr(address);
+
+    match state.0.load_session(address) {
+        Ok(Some(SerializedSession {
+            session,
+            extra_data,
+        })) => {
+            *record = session.into_raw();
+            if let Some(extra_data) = extra_data {
+                *user_record = extra_data.into_raw();
+            }
+
+            1
+        },
+        Ok(None) => 0,
+        Err(e) => e.code(),
+    }
+}
+
+unsafe extern "C" fn get_sub_device_sessions_func(
+    sessions: *mut *mut sys::signal_int_list,
+    name: *const c_char,
+    name_len: usize,
+    user_data: *mut c_void,
+) -> c_int {
+    assert!(!sessions.is_null());
+    assert!(!name.is_null());
+    assert!(!user_data.is_null());
+
+    let state = &*(user_data as *const State);
+    let name = std::slice::from_raw_parts(name as *const _, name_len);
+
+    match state.0.get_sub_device_sessions(name) {
+        Ok(got) => unimplemented!(),
```

Comment:
we need to revisit that, i could help with if you want :)

Reply:
Yep, good point. I originally skipped it because I'm lazy and  we need to manually convert the `Vec` into signal's linked list type.

Reply:
Turns out this was super easy to do! They provide `signal_int_list_alloc()` and `signal_int_list_push_back()` functions for creating and adding to the list.

Reply:
Yeah, ![lol](https://i.giphy.com/media/NaboQwhxK3gMU/giphy.webp)